### PR TITLE
Azure Monitor: Add azure batch api fallback logic

### PR DESCRIPTION
**Pull request by alyssajoyner**
Should be merged after https://github.com/grafana/grafana/pull/123696 

**What this does**

When a `metrics:getBatch` (Metrics Batch API) request fails with a **retryable** error such as throttling (429), Azure's 529 metrics throttle, 5xx, or a network failure, this re-fetches that batch's resources via the **ARM `/batch` endpoint** (`management.azure.com/batch?api-version=2020-06-01`) instead of failing the query.

This fallback is part of the Metrics Batch API feature and runs automatically whenever the batch path is active — it's gated behind the existing experimental `azureMonitorBatchAPI` flag.

**How it works**

1. A batch fails with a retryable status (`isRetryableStatus`: `0`/`429`/`5xx`).
2. We rebuild per-resource `/metrics` GET sub-requests **for that failed batch only** — reusing the existing `buildQuery`/`fanOutByResource` code so the URLs and params are identical to the legacy path (nothing is hand-reconstructed).
3. Those sub-requests are sent to ARM `/batch`, chunked to ≤20 per call to stay on the synchronous path.
4. Responses are parsed through the same `parseResponse` the single-resource path uses; per-resource failures are surfaced per-RefID while successful resources still return frames.

**Notes**

- **Only retryable failures trigger it.** Client errors (e.g. 400) are surfaced as-is.
- **If the ARM `/batch` call itself fails, the query fails.** 
- Uses the **ARM-audience client** against `management.azure.com`, not the metrics data-plane client.

### CODE REVIEW

**Review by wahajahmed010 — COMMENTED** [review]
## Review: Azure Monitor: Add azure batch api fallback logic

**Overall:** Solid implementation. The ARM /batch fallback is well-structured with clear separation of concerns. A few things worth noting:

### ✅ Good
- **Chunking logic** (`chunkSubRequests` at `maxARMBatchSize=20`) correctly avoids the async 202 path — smart.
- **`isRetryableStatus`** correctly handles 0 (network error), 429, and 5xx. No false positives for 4xx.
- **Test coverage is excellent** — retryable/non-retryable, chunking, partial failures, ARM /batch failure. The `TestExecuteBatchTimeSeriesQueryFallback` suite covers the edge cases well.
- **Reuses existing `parseResponse`** and `buildQuery` — no duplicated parsing logic.
- **`appendFrames`** is a nice small helper that cleans up the main flow.

### ⚠️ Concerns
1. **`parseFallbackResponse` error handling**: When a sub-response fails AND `parseResponse` also fails, the error from `parseFallbackResponse` is joined with `parseErr` from the chunk, but `appendFrames` already ran. The caller in `fallbackBatch` then appends `parseErr` to *all* queries in the chunk via `attachErr(result, q.RefID, parseErr)`. This means a resource that succeeded in `appendFrames` also gets an error attached. The frames are there but the error field is set — Grafana may show an error banner even though some data loaded. Consider only attaching errors to the specific failed sub-requests.

2. **`buildFallbackSubRequests`** iterates `batch.Queries` and calls `fanOutByResource` + `buildQuery` for each. If a query has many resources, this rebuilds the same query N times. Minor perf concern but worth noting for large dashboards.

3. **`maxBatchResponseBodyBytes`** is referenced but not defined in this file — ensure it's defined/imported correctly.

### 🔒 Security
- No injection vectors; all URL construction uses `strings.TrimRight` + `fmt.Sprintf` with validated params.
- Response body is `io.LimitReader` bounded.

**Verdict:** Approve with minor concern on error attribution in partial failures.

File: pkg/tsdb/azuremonitor/metrics/arm-batch-fallback.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+	"github.com/grafana/grafana/pkg/tsdb/azuremonitor/types"
+	"github.com/grafana/grafana/pkg/tsdb/azuremonitor/utils"
+)
+
+// The ARM /batch endpoint bundles many ARM reads into one request. It is undocumented and
+// used internally by the Azure Portal; the request/response contract below was verified
+// against live portal traffic. Because it has no official spec, it is gated behind a flag.
+// We use it as a fallback when the documented metrics:getBatch data-plane API fails with a
+// retryable error, wrapping each per-resource /metrics GET call as a sub-request.
```
Comment (aangelisc) [inline pkg/tsdb/azuremonitor/metrics/arm-batch-fallback.go]: Simplify this to just say the Azure Portal batch endpoint is the fallback. This is not gated by a flag.

File: pkg/tsdb/azuremonitor/metrics/arm-batch-fallback.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+	Content        json.RawMessage `json:"content"`
+}
+
+type armBatchResponseBody struct {
+	Responses []armBatchSubResponse `json:"responses"`
+}
+
+// executeARMBatch POSTs a set of sub-requests to the ARM /batch endpoint and returns the
+// parsed per-sub-request responses. Callers must keep len(subRequests) within
+// maxARMBatchSize to avoid the async path.
```
Comment (aangelisc) [inline pkg/tsdb/azuremonitor/metrics/arm-batch-fallback.go]: nit: unneeded comment

File: pkg/tsdb/azuremonitor/metrics/arm-batch-fallback.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+			}
+			for resourceURI := range azureQuery.Resources {
+				key := strings.ToLower(resourceURI)
+				if !wanted[key] || seen[key] {
+					continue
+				}
+				seen[key] = true
+				// GUID correlation name; keep the per-resource query so the
+				// response can be parsed with the existing parseResponse.
+				name := uuid.NewString()
```
Comment (aangelisc) [inline pkg/tsdb/azuremonitor/metrics/arm-batch-fallback.go]: nit: can we just use `refId` here?

**Review by aangelisc — DISMISSED** [review]
Generally LGTM, nice work on this! I left some nits for you 😊
